go/vt/vtgate/buffer/buffer.go

```diff
@@ -0,0 +1,244 @@
+package buffer
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	log "github.com/golang/glog"
+
+	"github.com/youtube/vitess/go/sync2"
+	"github.com/youtube/vitess/go/vt/discovery"
+	"github.com/youtube/vitess/go/vt/logutil"
+	"github.com/youtube/vitess/go/vt/topo/topoproto"
+	"github.com/youtube/vitess/go/vt/vterrors"
+
+	topodatapb "github.com/youtube/vitess/go/vt/proto/topodata"
+	vtrpcpb "github.com/youtube/vitess/go/vt/proto/vtrpc"
+)
+
+// Buffer is used to track ongoing MASTER tablet failovers and buffer
+// requests while the MASTER tablet is unavailable.
+// Once the new MASTER starts accepting requests, buffering stops and requests
+// queued so far will be automatically retried.
+//
+// There should be exactly one instance of this buffer. For each failover, an
+// instance of "ShardBuffer" will be created.
+type Buffer struct {
+	// Immutable configuration fields (parsed from command line flags).
+	// shards is a set of keyspace/shard entries to which buffering is limited.
+	// If empty (and *enabled==true), buffering is enabled for all shards.
+	shards map[string]bool
+	// shardsDryRun is the set of keyspace/shard entries for which dry-run mode
+	// should be enabled (failover tracking), even if buffering is disabled.
+	shardsDryRun map[string]bool
+
+	// bufferSizeSema limits how many requests can be buffered
+	// ("-vtgate_buffer_size") and is shared by all shardBuffer instances.
+	bufferSizeSema *sync2.Semaphore
+
+	// mu guards all fields in this group.
+	// In particular, it is used to serialize the following Go routines:
+	// - 1. Requests which may buffer (RLock, can be run in parallel)
+	// - 2. Request which starts buffering (based on the seen error)
+	// - 3. HealthCheck listener ("StatsUpdate") which stops buffering
+	// - 4. Timer which may stop buffering after -vtgate_buffer_max_failover_duration
+	mu sync.RWMutex
+	// failovers is the list of ongoing failovers.
+	// Key Format: "<keyspace>/<shard>"
+	failovers map[string]*shardBuffer
```

Comment:
Done.

I've changed the code such that the state is kept in `shardBuffer` now and these objects are reused across failovers.